salt/fileclient.py

```diff
@@ -1035,60 +1054,57 @@ def get_file(self,
             else:
-                hash_local = self.hash_file(dest2check, saltenv)
-                mode_local = None
-
-            if hash_local == hash_server:
-                if not salt.utils.is_windows():
-                    if mode_server is None:
-                        log.debug('No file mode available for \'%s\'', path)
-                    elif mode_local is None:
-                        log.debug(
-                            'No file mode available for \'%s\'',
-                            dest2check
-                        )
-                    else:
-                        if mode_server == mode_local:
-                            log.info(
-                                'Fetching file from saltenv \'%s\', '
-                                '** skipped ** latest already in cache '
-                                '\'%s\', mode up-to-date', saltenv, path
-                            )
-                        else:
-                            try:
-                                os.chmod(dest2check, mode_server)
-                                log.info(
-                                    'Fetching file from saltenv \'%s\', '
-                                    '** updated ** latest already in cache, '
-                                    '\'%s\', mode updated from %s to %s',
-                                    saltenv,
-                                    path,
-                                    salt.utils.st_mode_to_octal(mode_local),
-                                    salt.utils.st_mode_to_octal(mode_server)
-                                )
-                            except OSError as exc:
-                                log.warning(
-                                    'Failed to chmod %s: %s', dest2check, exc
-                                )
-                    # We may not have been able to check/set the mode, but we
-                    # don't want to re-download the file because of a failure
-                    # in mode checking. Return the cached path.
-                    return dest2check
-                else:
+                stat_local, hash_local = self.hash_and_stat_file(dest2check, saltenv)
+
+            mode_local = stat_local[0]
+            mtime_local = stat_local[8]
+            size_local = stat_local[6]
+
+            if support_mtime:
+                # Compare dates of modification and size
+                mtime_delta = mtime_server - mtime_local # if delta is > 0 then the file on the server was updated
+                size_ok = ( size_local == size_server )
+                log.debug("File sizes server/client: {0}, {1}, mtime delta = {2}".format(size_local, size_server, mtime_delta))
+
+                ok = mtime_delta <= 0 and size_ok # this is similar to how rsync would work
+            else:
+                # Compare hashes
+                log.debug("File hashes server/client {0}, {1}", hash_server, hash_local)
+                ok = ( hash_server == hash_local )
```

Comment:
Even if that is the case, it's not a great idea to have code that might hit a local that is only conditionally declared. It leaves the code prone to future regressions if it isn't carefully edited.

At minimum, it should at least be declared, even if it's set to ``None``.